tear down (#111)
cdr/docs: setup/teardown.md

---
title: Teardown
description: Learn how to teardown Coder and its associated infrastructure.
---

This guide shows you how to teardown Coder and the cluster it is deployed on.

**Note:** these instructions help you remove infrastructure spun up via our
[Kubernetes Documentation](https://coder.com/docs/setup/kubernetes), and do not
include teardown steps for any additional resources that may have been created

> outside of these steps

Reviewer: That being said, if there are other resources on a cluster, they will likely be deleted as well. 

Coder's recommended install path involved Coder being installed on to its own separate namespace. Perhaps the teardown docs could discuss how to delete all of the _Coder_ resources, and then separately how to destroy the cluster too?

Author: Good point - 

> > received the following:
> > `yarn run v1.22.10 warning package.json: No license field error Command "all-contributors" not found.`
> 
> Might mean you haven't run `yarn` to install?
> 
> I just performed the actions as a check:
> 
> ```
> greyb ~/Code/docs (vapurrmaid/chore-contributors) $ git checkout ericpaulsen/teardown 
> Branch 'ericpaulsen/teardown' set up to track remote branch 'ericpaulsen/teardown' from 'origin'.
> Switched to a new branch 'ericpaulsen/teardown'
> greyb ~/Code/docs (ericpaulsen/teardown) $ yarn
> yarn install v1.22.5
> warning package.json: No license field
> warning No license field
> [1/4] Resolving packages...
> [2/4] Fetching packages...
> [3/4] Linking dependencies...
> [4/4] Building fresh packages...
> 
> Done in 0.35s.
> greyb ~/Code/docs (ericpaulsen/teardown) $ yarn all-contributors add ericpaulsen content
> yarn run v1.22.5
> warning package.json: No license field
> $ /home/greyb/Code/docs/node_modules/.bin/all-contributors add ericpaulsen content
> Done in 0.28s.
> ```
> 
> Don't sweat it if it doesn't work - we'll have the all contributors bot up soon (thank you @bpmct). We'll make sure you get recorded!

thank you grey! I ran `git pull` to make sure I had the latest updates, for some reason I'm still receiving the error `Command "all-contributors" not found`.

Author: Understood - so this would involve adding steps to delete the namespace before the cluster delete commands are run. @khorne3 what are your thoughts?

Reviewer 2: That's not a bad idea -- I think it betters match our happy path install (but in reverse obviously)

Author: so I've just spoken with @sreya - we actually do not have a way to delete all coder-related resources in the cluster, while keeping other unrelated resources intact (but it is possible via a script). if we were to just delete the namespace, it wouldn't remote some cluster-scoped resources associated with coder, which would actually cause a reinstall to fail.

so for now, an alternative to simply deleting the whole cluster is to run `helm uninstall coder`, which will delete the coder services but keep the user environments (and their related disk space) intact.

what are the team's thoughts? @bpmct @khorne3

Reviewer: Thanks for the context. With that being said, I think that we can keep these instructions and then potentially add the script if/when is is ready.

I do think it's important to mention that deleting this teardown guide removes the whole cluster, which may potentially contain non-coder applications if the user installed extra stuff.

Author: Agreed - we do have language in this doc that says `This guide shows you how to tear down Coder and the cluster it is deployed on.` So that should be pretty clear to the user. 

As for the `helm uninstall` route, we can add a note saying that it will _only_ delete Coder `services` within the cluster, and all user environments will persist.

Reviewer 2: Can we adjust the callout to the following and if anyone asks, note that anything more will require some type of script? I'm afraid of spending a lot of time on something that will need modifications more often than not

> These instructions help you remove infrastructure created when following our
[Kubernetes setup tutorials](../setup/kubernetes/index.md). They do not include
teardown steps for any additional resources that you create. If you need to keep
your cluster, you can run `helm uninstall coder`, which deletes all Coder
services but retains environments and their associated disk space.

Author: That works well. I'm good with the changes above.

---
title: Teardown
description: Learn how to teardown Coder and its associated infrastructure.
---

This guide shows you how to teardown Coder and the cluster it is deployed on.

**Note:** these instructions help you remove infrastructure spun up via our
[Kubernetes Documentation](https://coder.com/docs/setup/kubernetes), and do not

> include teardown steps for any additional cluster resources that may have been deployed.

Reviewer 3: ```suggestion
include teardown steps for any additional resources that may have been created
outside of these steps
```

Just a suggestion, but it's not necessarily cluster resources. Things like IAM settings and other network security items.

Author: Makes sense - I'll include.

## Amazon Elastic Kubernetes Service (EKS)

1. Make sure you're running `eksctl` version 0.37.0 or later by running
the following command:

```bash
eksctl version
```

2. List all services in the cluster by running the following command:

```bash
kubectl get svc --all-namespaces
```

> 3. Delete any services that have an `EXTERNAL-IP` value
> by running the following command:
> 
> ```bash
> kubectl delete svc <service-name>
> ```

Reviewer 3: This needs an additional step:

You need to ensure `kubectl` is pointed to the same context still. Add steps to check and set the context to the right cluster before you run the delete command, or at least info to make sure people aren't blinding deleting services.

Maybe something like this would work:

```suggestion
3. Delete any services that have an `EXTERNAL-IP` value
in your namespace by running the following command:

```bash
kubectl delete svc <service-name>
```
```

4. Delete the cluster and its underlying nodes by
running the following command:

```bash
eksctl delete cluster --name <prod>
```

## Azure Kubernetes Serivce (AKS)

1. Ensure the environment variable for `RESOURCE_GROUP` is set
to the one you want to delete in Azure by running the following command:

```bash

> echo $RESOURCE_GROUP

Reviewer 3: ```suggestion
RESOURCE_GROUP="<MY_RESOURCE_GROUP>"
```

This makes more sense to me, though we may want to change the wording above instead or in addition to. This suggestions sets the variable and is explicit regardless of if you've lost that session in your terminal at some point.

The way it's worded with `echo`, would just show if a value is set. Maybe including the other part of the doc you got this from that states to set the value if it's not set?

Author: Yes - I left out the steps to set the value if it's not set, to try to keep things brief. However, I think over communicating a step or two will help make things clearer and facilitate more self-service.

```

2. Delete the cluster by running the following command:

```bash
az group delete --resource-group $RESOURCE_GROUP
```

## Google Kubernetes Engine (GKE)

1. Ensure the environment variables for `PROJECT_ID` and `CLUSTER_NAME` are still
set to the ones you want to delete in Google Cloud Platform (GCP).
Do this by running the following commands:

```bash
echo $PROJECT_ID
echo $CLUSTER_NAME

> ```

Reviewer 3: Similar to the Azure comment, may be worth also providing the commands to set the values or changing this to "set the values" instead of "check the values"

Author: Noted - will add this in.

2. Delete the cluster by running the following command:

> ```bash
> gcloud beta container --project $PROJECT_ID clusters delete 
> $CLUSTER_NAME --zone <enterzone>
> ```

Reviewer 3: ```suggestion
```bash
gcloud beta container --project $PROJECT_ID clusters delete \
$CLUSTER_NAME --zone <zone>
```
```

You just add a `\` to make multi-line commands on command line.

> Thank you for using Coder! If you would like to provide feedback, you can do so here.

Reviewer 3: I don't think this should be here. This teardown is something that could be literally just reprovisioning infrastructure to their trial ending. I don't think this feedback here (versus other existing mechanisms) makes as much sense, but open to others' thoughts.
